Merge conflicts: Only list files in worktree

gerhardol opened the pull request:
Contributes to #10423 

## Proposed changes

The "change" in index is a status marker only, there are no actual changes in the file.
The two appearances are confusing. It is required for a proper fix for #10423 
as the worktree status must be reset to HEAD, not the empty index.

Similar appearance changes in RevDiff and FormCommit.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/6248932/229382153-8221daba-6212-4e4f-b719-60d8fc598fb3.png)

![image](https://user-images.githubusercontent.com/6248932/229382168-7c238ce2-cac0-4eb8-8185-dba5caf3bb8e.png)

### After

![image](https://user-images.githubusercontent.com/6248932/229382298-436fd59d-b6e2-4b88-90c8-d79dfb59f94a.png)

## Test methodology <!-- How did you ensure quality? -->

Manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie reviewed (commented):
Do we have tests for this?

gerhardol commented:
> Do we have tests for this?

No, this is Git behavior so there is not much.
There is maybe something for the parsing of the status characters.

gerhardol commented:
Added a parsing test, including cases that do not fully work resetting
The cases with added files are not handled
https://git-scm.com/docs/git-status#_short_format

Left for later improvements...
![remote_deleted](https://user-images.githubusercontent.com/6248932/229901351-5cdda9cc-1a1d-4ddd-957a-a625dbb42d97.png)
![local_deleted](https://user-images.githubusercontent.com/6248932/229901366-5212f8c8-814f-4da9-86e5-bc0df7105342.png)

gerhardol commented:
For discussion, not this PR.

For the related conflicts where files have been deleted in any of the branches, the two commits provide information. For example an example where a file has been modified in both branches, deleted in the remote:

![image](https://user-images.githubusercontent.com/6248932/230502382-c7beae5a-4fda-4b3f-bb32-4c4630bc305c.png)

![image](https://user-images.githubusercontent.com/6248932/230502442-e0238734-3352-4c29-931b-11f5415026dd.png)

The worktree icon need to be changed to show both conflict and deleted in worktree, then the command to reset etc can be adjusted similar to #10854

mstv commented on GitCommands/Git/GetAllChangedFilesOutputParser.cs:
```diff
@@ -226,7 +227,12 @@ private static IReadOnlyList<GitItemStatus> GetAllChangedFilesFromString_v2(stri
 
-                    UpdateItemStatus(x, true, subm, fileName, oldFileName, renamePercent);
+                    // Skip unmerged where both are modified: Only worktree interesting.
+                    if (entryType != 'u' || x != 'U' || y != 'U')
```
ditto

gerhardol commented on GitCommands/Git/GetAllChangedFilesOutputParser.cs:
```diff
@@ -100,7 +100,8 @@ internal List<GitItemStatus> GetAllChangedFilesFromString_v1(string getAllChange
 
-                if (x != '?' && x != '!' && x != ' ')
+                // Skip unmerged where both are modified: Only worktree interesting.
+                if ((x != '?' && x != '!' && x != ' ') || !(x == 'U' && y == 'U'))
```
That was a a bigger change, to be consistent
Also modified the checks slightly, still ignoring unexpected data, required to ignore error printouts (this can maybe be changed when error output is no longer parsed, separate change).

mstv replied:
Thank you.

mstv commented on GitCommands/Git/GetAllChangedFilesOutputParser.cs:
```diff
@@ -175,67 +177,87 @@ private static IReadOnlyList<GitItemStatus> GetAllChangedFilesFromString_v2(stri
+                const char RenamedEntry = '2';
+                const char UnmergedEntry = 'u';
+
+                if (entryType != OrdinaryEntry && entryType != RenamedEntry && entryType != UnmergedEntry)
```
```suggestion
                if ((entryType != OrdinaryEntry && entryType != RenamedEntry && entryType != UnmergedEntry) || line.Length <= 3)
```

mstv commented on GitCommands/Git/GetAllChangedFilesOutputParser.cs:
```diff
@@ -92,19 +92,21 @@ internal List<GitItemStatus> GetAllChangedFilesFromString_v1(string getAllChange
-                if (x != '?' && x != '!' && x != ' ')
+                // Skip unmerged where both are modified: Only worktree interesting.
+                if ((x != GitItemStatusConverter.UntrackedStatus && x != GitItemStatusConverter.IgnoredStatus && x != GitItemStatusConverter.UnmodifiedStatus_v1)
+                    || !(x == GitItemStatusConverter.UnmergedStatus && y == GitItemStatusConverter.UnmergedStatus))
```
more similar to the change at line 248
```suggestion
                    || x != GitItemStatusConverter.UnmergedStatus || y != GitItemStatusConverter.UnmergedStatus)
```

gerhardol commented:
> The worktree icon need to be changed to show both conflict and deleted in worktree, then the command to reset etc can be adjusted similar to https://github.com/gitextensions/gitextensions/pull/10854

GitItemStatus need to keep information about both index/worktree to handle this correctly. One or two icons is maybe separate.

Will merge this, to be included in 4.1

RussKie commented on GitCommands/Git/GitItemStatusConverter.cs:
```diff
@@ -2,21 +2,33 @@
 {
     public static class GitItemStatusConverter
     {
+        public static readonly char AddedStatus = 'A';
+        public static readonly char CopiedStatus = 'C';
+        public static readonly char DeletedStatus = 'D';
+        public static readonly char ModifiedStatus = 'M';
+        public static readonly char IgnoredStatus = '!';
+        public static readonly char RenamedStatus = 'R';
+        public static readonly char SkippedStatus = 'S';
+        public static readonly char UnmergedStatus = 'U';
+        public static readonly char UnmodifiedStatus_v1 = ' ';
+        public static readonly char UnmodifiedStatus_v2 = '.';
+        public static readonly char UntrackedStatus = '?';
```
I think these can be turned into consts...

gerhardol replied:
It will not be a practical problem, the const will not be used in other assemblies. But this is the case with all except the plugin interface.
I set this as `static readonly` to not have to make changes, will be happy to change to `const` with comments...

RussKie commented:
If these are git-originated identifiers then we can presume these to be
constants, i think.

This is a non-blocking comment anyway.

On Tue, 11 Apr 2023, 10:09 pm Gerhard Olsson, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitCommands/Git/GitItemStatusConverter.cs
> <https://github.com/gitextensions/gitextensions/pull/10852#discussion_r1162724300>
> :
>
> > +        public static readonly char AddedStatus = 'A';
> +        public static readonly char CopiedStatus = 'C';
> +        public static readonly char DeletedStatus = 'D';
> +        public static readonly char ModifiedStatus = 'M';
> +        public static readonly char IgnoredStatus = '!';
> +        public static readonly char RenamedStatus = 'R';
> +        public static readonly char SkippedStatus = 'S';
> +        public static readonly char UnmergedStatus = 'U';
> +        public static readonly char UnmodifiedStatus_v1 = ' ';
> +        public static readonly char UnmodifiedStatus_v2 = '.';
> +        public static readonly char UntrackedStatus = '?';
>
> It will not be a practical problem, the const will not be used in other
> assemblies. But this is the case with all except the plugin interface.
> I set this as static readonly to not have to make changes, will be happy
> to change to const with comments...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/10852#discussion_r1162724300>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVZO2JSCBJRG547ET3XAVCZLANCNFSM6AAAAAAWQSIJHQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gerhardol commented:
Two minor changes when looking at the changes, merging in a couple of hours.
